src/bootstrap/compile.rs

```diff
@@ -883,16 +883,39 @@ impl Step for Rustc {
             compiler.host,
             target,
         );
+        let stamp = librustc_stamp(builder, compiler, target);
         run_cargo(
             builder,
             cargo,
             vec![],
-            &librustc_stamp(builder, compiler, target),
+            &stamp,
             vec![],
             false,
             true, // Only ship rustc_driver.so and .rmeta files, not all intermediate .rlib files.
         );
 
+        // When building `librustc_driver.so` (like `libLLVM.so`) on linux, it can contain
+        // unexpected debuginfo from dependencies, for example from the C++ standard library used in
+        // our LLVM wrapper. Unless we're explicitly requesting `librustc_driver` to be built with
+        // debuginfo (via the debuginfo level of the executables using it): strip this debuginfo
+        // away after the fact. This is to make the distributed artifacts smaller, and therefore we
+        // only do this at stage >= 1.
+        // FIXME: to make things simpler for now, limit this to the host and target where we know
+        // `strip -g` is both available and will fix the issue, i.e. on a x64 linux host that is not
+        // cross-compiling. Expand this to other appropriate targets in the future.
+        if compiler.stage != 0
+            && builder.config.rust_debuginfo_level_rustc == DebuginfoLevel::None
+            && builder.config.rust_debuginfo_level_tools == DebuginfoLevel::None
+            && target == "x86_64-unknown-linux-gnu"
+            && target == builder.config.build
+        {
+            let target_root_dir = stamp.parent().unwrap();
+            let rustc_driver = target_root_dir.join("librustc_driver.so");
+            if rustc_driver.exists() {
+                output(Command::new("strip").arg("--strip-debug").arg(rustc_driver));
+            }
+        }
```

Comment:
Checking the stage seems like unnecessary here. Isn't it already enough with the condition on whether debug-info was requested or not?

Reply:
The config (and thus the debuginfo request) will be the same for all stages currently, or not?

Reply:
> The config (and thus the debuginfo request) will be the same for all stages currently, or not?

Currently, will be same only if stage is not 0. Which is the reason why I am offering to remove that condition.

Reply:
I don't believe there's consistency when compiling librustc_driver.so at different stages, but sure, I've pushed a commit to remove the check.